### PR TITLE
parsing: Support dict unpacking in `cmd`.

### LINKED ISSUE #6107: More flexible dvc.yaml parameterisation
## Context

One of DVC usage scenarios we should consider is the following:
1. User has a big and complex real-life repo and workflows around it (one example is CI/CD)
1. User wants to apply DVC without modifications to his current scripts, because that will trigger code changes in subsequent workflows, which the user don’t want to do right now, but rather later, when DVC will be integrated and will prove it is useful for the user.

One example for this is the [YOLOv5 repo](https://github.com/ultralytics/yolov5). The github repo has only [testing in CI](https://github.com/ultralytics/yolov5/blob/develop/.github/workflows), but for sure we can have a private fork and multiple CI/CD workflows which train, test, export and deploy models.

## Problem

In this case user could have scripts [which run like this](https://github.com/ultralytics/yolov5/blob/develop/.github/workflows/ci-testing.yml#L67):
```
python train.py --img 128 --batch 16 --weights weights/${{ matrix.model }}.pt --cfg models/${{ matrix.model }}.yaml --epochs 1 --device $di
```

The scripts could have a lot of args (this `train.py` has about 20 of them, all handled by Python argparse module). If I don’t want to modify my scripts yet, I’d need to duplicate all arguments in two places: in params.yaml and in dvc.yaml, thus having three slightly different copies of the arguments.

Indeed, this is a bad practice that creates a technical debt of maintaining three copies of the same and shouldn’t be used for too long, but as a user I would prefer to migrate to DVC step by step, not breaking anything in current workflows in the process. And if I’m not sure I want to migrate and just want to give DVC a try I would want to keep all my current scripts unchanged to remove DVC later if I won’t be planning on using it.

## Suggestion
Expand args dictionary and lists from params in dvc.yaml to allow user to avoid duplicating every param in dvc.yaml:
```
# params.yaml
train-params:
  param1: 1
  param2: 2
  param3: [value1, value2]


# dvc.yaml
stages:
  train:
    cmd: python train.py ${train-params}
```

which gets translated to
```
python train.py --param1 1 --param2 2 --param3 value1 value2
```

## Few questions we should also consider:

<details>
<summary><b>1. Is there a use case when a user wants to expand a nested dict?</b></summary>

Example:
```
# params.yaml
train-params:
  param1:
    nested-param: value

# dvc.yaml
stages:
  train:
    cmd: python train.py ${train-params}
```
</details>

<details>
<summary><b>2. There are optional args like `--resume` which could be supplied or not. We could support this like following:</b></summary>

```
# dvc.yaml
stages:
  train:
    deps:
      - train.py
    cmd: python train.py --resume ${resume}
```

In case we want to run just “python train.py”
```
$ dvc exp run -S resume= 
> python train.py

# params.yaml
resume:
```
In case we want to run “python train.py --resume path-to-weights.pt”:
```
$ dvc exp run -S resume=path-to-weights.pt
> python train.py --resume path-to-weights.pt

# params.yaml
resume: path-to-weights.pt
```
Right now the first use case is not supported at all:
```
$ dvc exp run -S param=  
ERROR: Must provide a value for parameter 'param'
```
Though params.yaml could be like
```
# params.yaml
param:
```
And experiment could be run, but “None” value will be used
```
$ dvc exp run
> python train.py None
```
That said, the last modification could break backwards compatibility if anyone uses the later approach with None substitution, but it looks more like a bug, then a solution, and we can replace this behaviour in the future.
</details>

<details>
<summary><b>3. Are there other examples of CLI parameters that scripts can take, but dvc.yaml / `dvc exp run` doesn't support now?</b></summary>

Would be great to gather them is well.
</details>

My perspective here is limited to python and bash scripts, so I can miss something important. I'm going to gather more examples and possible issues and add more information about them. Meanwhile, it would be great to hear what others think.

### PR DESCRIPTION
**Pull request by daavoo**
Allow to use dictionaries as values for template interpolation but only inside the `cmd` key.

Given the following params.yaml and dvc.yaml:

```yaml
# params.yaml
dict:
  foo: foo
  bar: bar
```

```yaml
# dvc.yaml
stages:
  stage1:
    cmd: python script.py ${dict}
```

The dictionary will be unpacked with the following syntax:

```yaml
# dvc.yaml
stages:
  stage1:
    cmd: python script.py --foo foo --bar bar
```

Closes #6107

* [x] ❗ I have followed the [Contributing to DVC](https://dvc.org/doc/user-guide/contributing/core) checklist.

* [ ] 📖 If this PR requires [documentation](https://dvc.org/doc) updates, I have created a separate PR (or issue, at least) in [dvc.org](https://github.com/iterative/dvc.org) and linked it here.

### CODE REVIEW

**Comment by skshetry** [comment]
> Allow to use dictionaries as values for template interpolation but only inside the `cmd` key.
> 
> Given the following params.yaml and dvc.yaml:
> 
> ```yaml
> # params.yaml
> dict:
>   foo: foo
>   bar: bar
> ```
> 
>  ```yaml
> # dvc.yaml
> stages:
>   stage1:
>     cmd: python script.py ${dict}
> ```
> 
> The dictionary will be unpacked with the following syntax:
> 
> ```yaml
> # dvc.yaml
> stages:
>   stage1:
>     cmd: python script.py --foo foo --bar bar
> ```
> 
> Closes #6107
> 
> * [x]  ❗ I have followed the [Contributing to DVC](https://dvc.org/doc/user-guide/contributing/core) checklist.
> * [ ]  📖 If this PR requires [documentation](https://dvc.org/doc) updates, I have created a separate PR (or issue, at least) in [dvc.org](https://github.com/iterative/dvc.org) and linked it here.

Any recent motivations for this? 🙂

**Comment by daavoo** [comment]
> Any recent motivations for this? 🙂

When exploring how to better integrate `dvc params` with ML frameworks, this showed as low-hanging fruit for some frameworks like [Yolov5](https://github.com/iterative/yolov5) (linked issue), [HuggingFace](https://huggingface.co/docs/transformers/v4.20.0/en/internal/trainer_utils#transformers.HfArgumentParser) , [PyTorch Lightning](https://pytorch-lightning.readthedocs.io/en/stable/common/lightning_cli.html#use-of-command-line-arguments), etc. 
In addition, the default syntax is good enough for custom training scripts using simple `argparse` and having more flexibility of configuring `cmd` calls of CLI tools without having to edit `dvc.yaml` for each iteration (adding/removing argument)

File: tests/func/parsing/test_interpolated_entry.py [hunk 1]
```diff
@@ -259,3 +261,13 @@ def test_vars_load_partial(tmp_dir, dvc, local, vars_):
         d["vars"] = vars_
     resolver = DataResolver(dvc, tmp_dir.fs_path, d)
     resolver.resolve()
+
+
+def test_cmd_dict(tmp_dir, dvc):
+    (tmp_dir / DEFAULT_PARAMS_FILE).dump(CONTEXT_DATA)
```
Comment (dberenbaum) [inline tests/func/parsing/test_interpolated_entry.py:267]: What about non-default params files? Doesn't seem to work with those right now.

**Comment by dberenbaum** [comment]
Some data types seem like they get tricky 😓 :

* Strings: We probably need to quote these? Otherwise, they can easily turn into a mess like "--model random forest"
* Booleans: I think these should be expanded to just `--foo` with no value if true and ignored if false. Or they could be set to `--no-foo` if false, but this seems like too much magic to me. 
* Iterables: They don't work right now. I'd suggest `--foo 1 --foo 2` to be explicit and consistent with argparse.

File: tests/func/parsing/test_interpolated_entry.py (hunk 1 above)
Comment (daavoo) [reply tests/func/parsing/test_interpolated_entry.py:267]: > What about non-default params files? Doesn't seem to work with those right now.

Not sure I follow, the changes in this P.R. should work the same way as the current templating behavior

Comment (dberenbaum) [reply tests/func/parsing/test_interpolated_entry.py]: My misunderstanding

File: tests/func/parsing/test_interpolated_entry.py [hunk 2]
```diff
@@ -259,3 +259,38 @@ def test_vars_load_partial(tmp_dir, dvc, local, vars_):
         d["vars"] = vars_
     resolver = DataResolver(dvc, tmp_dir.fs_path, d)
     resolver.resolve()
+
+
+def test_cmd_dict(tmp_dir, dvc):
+    data = {
```
[the remark this answers is not in the document]
Comment (dberenbaum) [reply tests/func/parsing/test_interpolated_entry.py:265]: > I ignored it for now (nothing gets added to the interpolated string) .
> Maybe `--no-{x}` makes sense (i.e. is also the [default used in click](https://click.palletsprojects.com/en/8.1.x/options/#boolean-flags)) , the worst thing that can happen is that we will add a flag that it's ignored?

No strong preference here.

> Not sure if we should raise an explicit exception if we detect a nested list or just render as if it was a string, which will probably raise an exception on the user code when parsing the args.

If easy enough, I would prefer the former, but doesn't need to block the PR.

[the remark this answers is not in the document]
Comment (daavoo) [reply tests/func/parsing/test_interpolated_entry.py:265]: > No strong preference here.

Added the `--no-` prefix for negative booleans instead of ignoring them.

> If easy enough, I would prefer the former, but doesn't need to block the PR.

Added explicit exception

File: dvc/parsing/interpolate.py [hunk 3]
```diff
@@ -80,6 +81,31 @@ def _(obj: bool):
     return "true" if obj else "false"
 
 
+@to_str.register(dict)
+def _(obj: dict):
+    result = ""
+    for k, v in flatten(obj).items():
+        if isinstance(v, bool):
```
Comment (karajan1001) [inline dvc/parsing/interpolate.py:88]: This only works for `argparse.BooleanOptionalAction` not for `store_true/false`. Maybe need some better way to handle this?

File: dvc/parsing/interpolate.py [hunk 4]
```diff
@@ -80,6 +81,31 @@ def _(obj: bool):
...
+            result += f"--{k} '{v}' "
+        elif isinstance(v, Iterable):
+            for i in v:
+                if isinstance(i, str):
+                    i = f"'{i}'"
+                elif isinstance(i, Iterable):
+                    raise ParseError(
+                        f"Cannot interpolate nested iterable in '{k}'"
+                    )
+                result += f"--{k} {i} "
```
Comment (karajan1001) [inline dvc/parsing/interpolate.py:103]:
```python
> import argparse
> parser = argparse.ArgumentParser()
> parser.add_argument('--foo', nargs='*')
> parser.parse_args('--foo x y'.split())
Namespace(foo=['x', 'y'])

> parser.parse_args('--foo x --foo y'.split())
Namespace(foo=['y'])
```

Looks like setting arguments for multi times will overwrite the previous one.

File: dvc/parsing/interpolate.py (hunk 3 above)
Comment (daavoo) [reply dvc/parsing/interpolate.py:88]: The same issue about different options, as commented below https://github.com/iterative/dvc/pull/7907#discussion_r903429914 

Which option do you think we should consider as the most appropriate default? 

Note that this can be used for things beyond `argparse`, for example you can use the interpolation to pass flags to arbitrary executables that you call inside `cmd`

File: dvc/parsing/interpolate.py (hunk 4 above)
Comment (dberenbaum) [reply dvc/parsing/interpolate.py:103]: Let's pick one behavior for now. Is it possible to see which one is more widely used in existing ML CLIs? I would guess the `nargs` `--foo 1 2` behavior is more common, even though I find the explicit `append` behavior easier to read.

Comment (daavoo) [reply dvc/parsing/interpolate.py]: Added `config.parsing.list` with `nargs` and `append` options. It doesn't look that bad to have the options.

File: dvc/parsing/interpolate.py (hunk 3 above)
Comment (karajan1001) [reply dvc/parsing/interpolate.py:93]: I also have no idea about it.

Comment (dberenbaum) [reply dvc/parsing/interpolate.py:93]: I think that's more than enough to support for now.

File: tests/func/parsing/test_interpolated_entry.py [hunk 5]
```diff
@@ -259,3 +259,59 @@ def test_vars_load_partial(tmp_dir, dvc, local, vars_):
         d["vars"] = vars_
     resolver = DataResolver(dvc, tmp_dir.fs_path, d)
     resolver.resolve()
+
+
+@pytest.mark.parametrize(
+    "bool_config", [None, "store_true", "boolean_optional"]
```
Comment (karajan1001) [inline tests/func/parsing/test_interpolated_entry.py:265]: 1. Maybe we only need to test two different conditions? The `None` can be tested through some other method (Test if the default value is set corretly)?
2. Because these two configurations is independent, maybe we can test them in an (I'm not sure how to describe it) instead of using a product.  ["a1b1", "a2b2"] instead of ["a1", "a2"] x ["b1", "b2"]. 
The two suggestion above can reduce the tests amount from 9 to 2.

**Comment by karajan1001** [comment]
LGTM for the most part, another concern is that the two new configurations still can't solve all of the problems as two different types of` arg` might exist in one command.

File: tests/func/parsing/test_interpolated_entry.py (hunk 5 above)
Comment (daavoo) [reply tests/func/parsing/test_interpolated_entry.py]: I updated it to 3 test cases (`["a1b1", "a2b2"]`). I still included the default in the same test, not sure if it's worth it to separate it as the body of the test would be the same

**Comment by daavoo** [comment]
> another concern is that the two new configurations still can't solve all of the problems as two different types of arg might exist in one command.

Indeed. There are many unsolved issues but the idea for this initial P.R. is to cover basic scenarios to don't overcomplicate the logic until users request for specific issues

**Review by karajan1001 — APPROVED** [review]

File: dvc/parsing/interpolate.py (hunk 4 above)
Comment (skshetry) [reply dvc/parsing/interpolate.py]: Just noticed this, it's a bad idea to make `dvc.yaml` parsing conditional, it should be self-contained.
